src/charm.py

```diff
@@ -481,741 +43,243 @@ def __init__(self, *args) -> None:  # type: ignore[no-untyped-def]
+        """Create an ingress controller.
 
         Args:
-            event: Juju event that fires this handler.
-        """
-        api = self._networking_v1_api()
-        ingresses = api.list_namespaced_ingress(namespace=self._namespace)
-        event.set_results({"ingresses": ingresses})
-
-    def k8s_auth(self) -> None:
-        """Authenticate to kubernetes."""
-        if self._authed:
-            return
-
-        kubernetes.config.load_incluster_config()
-
-        self._authed = True
-
-    def _core_v1_api(self) -> kubernetes.client.CoreV1Api:
-        """Use the v1 k8s API.
+            namespace: Kubernetes namespace.
 
         Returns:
-            The core v1 API.
+            An IngressController instance.
         """
-        self.k8s_auth()
-        return kubernetes.client.CoreV1Api()
+        return IngressController(namespace=namespace, labels=self._labels)
 
-    def _networking_v1_api(self) -> kubernetes.client.NetworkingV1Api:
-        """Use the v1 beta1 networking API.
+    def _get_relation(self) -> Optional[Relation]:
+        """Get the current effective relation.
 
         Returns:
-            The networking v1 API.
-        """
-        self.k8s_auth()
-        return kubernetes.client.NetworkingV1Api()
+            The current effective relation object, None if there are no relation.
+        """
+        if self.model.get_relation("nginx-route") is not None:
+            relation = cast(Relation, self.model.get_relation("nginx-route"))
+            if relation.app is not None and relation.data[relation.app]:
+                return relation
+        elif self.model.get_relation("ingress") is not None:
+            relation = cast(Relation, self.model.get_relation("ingress"))
+            if relation.app is not None and relation.data[relation.app]:
+                return relation
+        return None
 
-    def _report_service_ips(self) -> List[str]:
-        """Report on service IP(s) and return a list of them.
+    def _get_definition_from_relation(self, relation: Relation) -> IngressDefinition:
+        """Get the IngressDefinition from the given relation.
 
-        Returns:
-            A list of service IPs.
+        Args:
+            relation: The source relation object.
+
+        Return:
+            The IngressDefinition corresponding to the provided relation.
         """
-        api = self._core_v1_api()
-        services = api.list_namespaced_service(  # type: ignore[attr-defined]
-            namespace=self._namespace
-        )
-        all_k8s_service_names = [rel._k8s_service_name for rel in self._all_config_or_relations]
-        return [
-            x.spec.cluster_ip for x in services.items if x.metadata.name in all_k8s_service_names
-        ]
+        if relation.name == "nginx-route":
+            definition_essence = IngressDefinitionEssence(
+                model=self.model, config=self.config, relation=relation
+            )
+        else:
```

Comment:
Please change this to check that the relation name is `ingress` and have an else condition that raises an error for defensive programming purposes

Reply:
Exhaustiveness checking added, thanks!